### ISSUE TITLE
Stack Overflow with String Sort Solve on String length 2916

### ISSUE THREAD

**Issue by agvishal477** [Z3Prover/z3#2850]
Hi

This is in continuation and related to issue #2821. 
After further investigation into the issue, I've found out that z3Solver.check() call fails at string_value length 2916 in version 4.8.4 and at 2565 in further version upto 4.8.7.
This happens when check() call is made between push() and pop() calls.
If push and pop are not used string value length even upto 15000 are supported.

Can you please explain what is causing this and confirm the supported length of string_val which can be safely used in the following case?

``` z3::context z3Context;
        z3::expr AFam = z3Context.constant("A", z3Context.string_sort());
        // AValue length is 2565 chars.
        z3::expr AValue = z3Context.string_val("LoremipsumdolorsitametconsecteturadipiscingelitseddoeiusmodtemporincididuntutlaboreetdoloremagnaaliquaUtenimadminimveniamquisnostrudexercitationullamcolaborisnisiutaliquipexeacommodoconsequatDuisauteiruredolorinreprehenderitinvoluptatevelitessecillumdoloreeufugiatnullapariaturExcepteursintoccaecatcupidatatnonproidentsuntinculpaquiofficiadeseruntmollitanimidestlaborumLoremipsumdolorsitametconsecteturadipiscingelitseddoeiusmodtemporincididuntutlaboreetdoloremagnaaliquaUtenimadminimveniamquisnostrudexercitationullamcolaborisnisiutaliquipexeacommodoconsequatDuisauteiruredolorinreprehenderitinvoluptatevelitessecillumdoloreeufugiatnullapariaturExcepteursintoccaecatcupidatatnonproidentsuntinculpaquiofficiadeseruntmollitanimidestlaborumLoremipsumdolorsitametconsecteturadipiscingelitseddoeiusmodtemporincididuntutlaboreetdoloremagnaaliquaUtenimadminimveniamquisnostrudexercitationullamcolaborisnisiutaliquipexeacommodoconsequatDuisauteiruredolorinreprehenderitinvoluptatevelitessecillumdoloreeufugiatnullapariaturExcepteursintoccaecatcupidatatnonproidentsuntinculpaquiofficiadeseruntmollitanimidestlaborumLoremipsumdolorsitametconsecteturadipiscingelitseddoeiusmodtemporincididuntutlaboreetdoloremagnaaliquaUtenimadminimveniamquisnostrudexercitationullamcolaborisnisiutaliquipexeacommodoconsequatDuisauteiruredolorinreprehenderitinvoluptatevelitessecillumdoloreeufugiatnullapariaturExcepteursintoccaecatcupidatatnonproidentsuntinculpaquiofficiadeseruntmollitanimidestlaborumLoremipsumdolorsitametconsecteturadipiscingelitseddoeiusmodtemporincididuntutlaboreetdoloremagnaaliquaUtenimadminimveniamquisnostrudexercitationullamcolaborisnisiutaliquipexeacommodoconsequatDuisauteiruredolorinreprehenderitinvoluptatevelitessecillumdoloreeufugiatnullapariaturExcepteursintoccaecatcupidatatnonproidentsuntinculpaquiofficiadeseruntmollitanimidestlaborumLoremipsumdolorsitametconsecteturadipiscingLoremipsumdolorsitametconsecteturadipiscingelitseddoeiusmodtemporincididuntutlaboreetdoloremagnaaliquaUtenimadminimveniamquisnostrudexercitationullamcolaborisnisiutaliquipexeacommodoconsequatDuisauteiruredolorinreprehenderitinvoluptatevelitessecillumdoloreeufugiatnullapariaturExcepteursintoccaecatcupidatatnonproidentsuntinculpaquiofficiadeseruntmollitanimidestlaborumLoremipsumdolorsitametconsecteturadipiscingelitseddoeiusmodtemporincididuntutlaboreetdoloremagnaaliquaUtenimadminimveniamquisnostrudexercitationullamcolaborisnisiutaliquipexeacommodoconsequatDuisauteiruredolorinreprehenderitinvoluptatevelitessecillumdoloreeufritinvoluptatevelitessecillumdoloreeufritinvabcde");
        z3::solver z3Solver(z3Context);

        z3Solver.push();
        z3Solver.add(AFam == AValue);
        z3::check_result verdict = z3Solver.check();
        std::cout << verdict << std::endl;
        if (z3::sat == verdict)
        {
            std::cout << z3Solver.get_model() << std::endl;
        }
        z3Solver.pop();
```

Thanks

**Comment by NikolajBjorner** [comment]
Fixed with ab5905cf7fd5299fa27d76ee2d3aa959d68f8ec0

**Comment by NikolajBjorner** [comment]
Perf is still not good: quadratic overhead during model building 
because it creates values for all prefixes. While the other perf bugs that were addressed transcend whether using long strings, I am sure how important 
the use case of long string is, though. So holding off from touching that part.
